### PR TITLE
spring mvc3 - proper span name in case of sec fail (#718)

### PR DESCRIPTION
**Pull request by kubawach**
- fixed the issue - caused by not allowing specific spring boot context implementation (AnnotationConfigServletWebServerApplicationContext) to be instrumented
- added test case for this particular issue
- switched test instrumentation to use spring boot 2 (de facto standard for many years now)
- light refactoring of the AdditionalLibraryIgnoresMatcher
- added instructions for running single test to MD file
- added few missing tests

### CODE REVIEW

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/main/java/io/opentelemetry/instrumentation/auto/springwebmvc/WebApplicationContextInstrumentation.java [hunk 1]
```diff
@@ -17,8 +17,7 @@
 package io.opentelemetry.instrumentation.auto.springwebmvc;
 
 import static io.opentelemetry.javaagent.tooling.ClassLoaderMatcher.hasClassesNamed;
-import static io.opentelemetry.javaagent.tooling.bytebuddy.matcher.AgentElementMatchers.extendsClass;
-import static io.opentelemetry.javaagent.tooling.bytebuddy.matcher.AgentElementMatchers.implementsInterface;
+import static io.opentelemetry.javaagent.tooling.bytebuddy.matcher.AgentElementMatchers.*;
```
Comment (mateuszrzeszutek) [inline instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/main/java/io/opentelemetry/instrumentation/auto/springwebmvc/WebApplicationContextInstrumentation.java:20]: Make sure to disable wildcard import in IntelliJ both for Java and Groovy

File: javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/tooling/matcher/AdditionalLibraryIgnoresMatcherTest.groovy [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    setup:
+    def type = Mock(TypeDescription)
+    type.getActualName() >> typeName
+
+    when:
+    def matches = underTest.matches(type)
+
+    then:
+    matches == true
```
Comment (mateuszrzeszutek) [inline javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/tooling/matcher/AdditionalLibraryIgnoresMatcherTest.groovy:20]: You can use `expect:` instead

```suggestion
    expect:
    underTest.matches(type)
```

File: docs/contributing/running-tests.md [hunk 3]
```diff
@@ -41,3 +41,7 @@ rule will kick in and do the following:
 This works both for tasks named `test` and `latestDepTest`. But currently
 does not work for other custom test tasks, such as those created by test sets
 plugin.
+
+#### Executing single test
+
+Executing `./gradlew :instrumentation:<INSTRUMNETATION_NAME>>:test --tests <GROOVY TEST FILE NAME>` will run only the selected test.
```
Comment (mateuszrzeszutek) [inline docs/contributing/running-tests.md:47]: Typo: `INSTRUMNETATION_NAME` -> `INSTRUMENTATION_NAME`

Also you have one `>` too many

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java [hunk 4]
```diff
@@ -254,13 +254,13 @@ public void onIgnored(
     @Override
     public void onComplete(
         String typeName, ClassLoader classLoader, JavaModule module, boolean loaded) {
-      //      log.debug("onComplete {}", typeName);
+        // log.debug("onComplete {}", typeName);
```
Comment (mateuszrzeszutek) [inline javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java:257]: Is this change needed?

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/groovy/test/filter/FilteredAppConfig.groovy [hunk 5]
```diff
@@ -16,6 +16,8 @@
 
 package test.filter
 
+import org.springframework.boot.web.embedded.tomcat.TomcatConnectorCustomizer
```
Comment (mateuszrzeszutek) [inline instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/groovy/test/filter/FilteredAppConfig.groovy:19]: Make sure that you have [correct import settings](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/docs/contributing/intellij-setup.md)

**Review by mateuszrzeszutek — APPROVED** [review]

**Comment by iNikem** [comment]
@kubawach Can you please rebase this PR and mark resolved comments as such?

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/main/java/io/opentelemetry/instrumentation/auto/springwebmvc/WebApplicationContextInstrumentation.java (hunk 1 above)
Comment (kubawach) [reply instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/main/java/io/opentelemetry/instrumentation/auto/springwebmvc/WebApplicationContextInstrumentation.java]: Corrected.

File: javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/tooling/matcher/AdditionalLibraryIgnoresMatcherTest.groovy (hunk 2 above)
Comment (kubawach) [reply javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/tooling/matcher/AdditionalLibraryIgnoresMatcherTest.groovy:36]: I like to stick to GWT BDD test convention.

File: docs/contributing/running-tests.md (hunk 3 above)
Comment (kubawach) [reply docs/contributing/running-tests.md]: Good catch!

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/groovy/test/filter/FilteredAppConfig.groovy (hunk 5 above)
Comment (kubawach) [reply instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/groovy/test/filter/FilteredAppConfig.groovy]: Yep, done.

**Comment by kubawach** [comment]
> @kubawach Can you please rebase this PR and mark resolved comments as such?

Done. I also managed to work around another Spring Boot 2 change failing one of our tests.

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/main/java/io/opentelemetry/instrumentation/auto/springwebmvc/WebApplicationContextInstrumentation.java [hunk 6]
```diff
@@ -17,15 +17,14 @@
 package io.opentelemetry.instrumentation.auto.springwebmvc;
 
 import static io.opentelemetry.javaagent.tooling.ClassLoaderMatcher.hasClassesNamed;
-import static io.opentelemetry.javaagent.tooling.bytebuddy.matcher.AgentElementMatchers.extendsClass;
```
Comment (iNikem) [inline instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/main/java/io/opentelemetry/instrumentation/auto/springwebmvc/WebApplicationContextInstrumentation.java:20]: Any reason for this change?

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/spring-webmvc-3.1-auto.gradle [hunk 7]
```diff
@@ -39,9 +39,9 @@ dependencies {
 
   testImplementation group: 'org.spockframework', name: 'spock-spring', version: "$versions.spock"
 
-  testImplementation group: 'org.springframework.boot', name: 'spring-boot-starter-test', version: '1.5.17.RELEASE'
-  testImplementation group: 'org.springframework.boot', name: 'spring-boot-starter-web', version: '1.5.17.RELEASE'
-  testImplementation group: 'org.springframework.boot', name: 'spring-boot-starter-security', version: '1.5.17.RELEASE'
+  testImplementation group: 'org.springframework.boot', name: 'spring-boot-starter-test', version: '2.3.1.RELEASE'
```
Comment (iNikem) [inline instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/spring-webmvc-3.1-auto.gradle:42]: I am not sure about this change. We usually test the oldest supported version by default and let `latestDepTests` verify the latest supported version.

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/resources/logback.xml [hunk 8]
```diff
@@ -14,5 +14,6 @@
   </root>
 
   <logger name="io.opentelemetry.auto" level="debug"/>
+  <logger name="io.opentelemetry.javaagent" level="debug"/>
```
Comment (iNikem) [inline instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/resources/logback.xml:17]: Remove that line above. I believe it is obsolete

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/main/java/io/opentelemetry/instrumentation/auto/springwebmvc/WebApplicationContextInstrumentation.java (hunk 6 above)
Comment (kubawach) [reply instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/main/java/io/opentelemetry/instrumentation/auto/springwebmvc/WebApplicationContextInstrumentation.java:20]: I blame intelliJ. Will revert.

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/resources/logback.xml (hunk 8 above)
Comment (kubawach) [reply instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/resources/logback.xml:17]: Done.

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/spring-webmvc-3.1-auto.gradle (hunk 7 above)
Comment (kubawach) [reply instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/spring-webmvc-3.1-auto.gradle:42]: Is it changed to 2.x stream by latestDep? I had to make a few changes around test code (AppConfig, ServletFIlteTest) to make it work (spring boot 2 is in some aspects not compatible with 1.x stream). I assumed it's worth upgrading (in baseline) as boot 2 has been around for more than 2 years now. Let me know what you think about that.

Comment (iNikem) [reply instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/spring-webmvc-3.1-auto.gradle:42]: I just realised that SB 2.3.1 most certainly does not use Spring MVC 3.1. So after these changes we don't test if we actually support 3.1. And as there are API changes affecting tests, it means we should test them separately but certainly should test them both. So I propose to leave this with old version as it was. And we should add a separate test with more recent versions of Spring MVC. You don't have to add that new test right away, you can add a follow up issue.

Comment (kubawach) [reply instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/spring-webmvc-3.1-auto.gradle]: Per slack conversation - made slight adjustments for the code to get both boot 1.x and 2.x compatibility.

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/groovy/test/filter/FilteredAppConfig.groovy [hunk 9]
```diff
@@ -53,6 +41,15 @@ import org.springframework.web.context.request.NativeWebRequest
 import org.springframework.web.servlet.config.annotation.ContentNegotiationConfigurer
 import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter
 
+import javax.servlet.Filter
```
Comment (iNikem) [inline instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/groovy/test/filter/FilteredAppConfig.groovy:44]: Please run `./gradlew spotlessApply` and configure your IDE as described [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/docs/contributing/intellij-setup.md)

**Review by iNikem — APPROVED** [review]

File: instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/groovy/test/filter/FilteredAppConfig.groovy (hunk 9 above)
Comment (kubawach) [reply instrumentation/spring/spring-webmvc-3.1/spring-webmvc-3.1-auto/src/test/groovy/test/filter/FilteredAppConfig.groovy:44]: done!
